cooperative_lanechange/src/cooperative_lanechange.cpp

```diff
@@ -0,0 +1,867 @@
+ 
+        }
+
+        return true;
+    }
+    
+    cav_msgs::MobilityRequest CooperativeLaneChangePlugin::create_mobility_request(std::vector<cav_msgs::TrajectoryPlanPoint>& trajectory_plan, cav_msgs::Maneuver maneuver){
+
+        cav_msgs::MobilityRequest request_msg;
+        cav_msgs::MobilityHeader header;
+        header.sender_id = sender_id_;
+        header.recipient_id = DEFAULT_STRING_;      //To do- find way to query recipient id
+        header.sender_bsm_id = bsmIDtoString(bsm_core_);
+        header.plan_id = boost::uuids::to_string(boost::uuids::random_generator()());
+        header.timestamp = trajectory_plan.front().target_time.toSec();
+        request_msg.header = header;
+
+        request_msg.strategy = "carma/cooperative-lane-change";
+        request_msg.plan_type.type = cav_msgs::PlanType::CHANGE_LANE_LEFT;
+        //Urgency- Currently unassigned
+
+        //Location
+        cav_msgs::LocationECEF location;
+        location.ecef_x = pose_msg_.pose.position.x;
+        location.ecef_y = pose_msg_.pose.position.y;
+        location.ecef_z = pose_msg_.pose.position.z;
+        //Using trajectory first point time as location timestamp;
+        location.timestamp = trajectory_plan.front().target_time.toSec();
+        request_msg.location = location;
+        //Strategy params
+        //Encode JSON with Boost Property Tree
+        using boost::property_tree::ptree;
+        ptree pt;
+        pt.put("speed",maneuver.lane_change_maneuver.start_speed);
+        pt.put("start_lanelet",maneuver.lane_change_maneuver.starting_lane_id);
+        pt.put("end_lanelet", maneuver.lane_change_maneuver.ending_lane_id);
+        std::stringstream body_stream;
+        boost::property_tree::json_parser::write_json(body_stream,pt);
+        request_msg.strategy_params = body_stream.str();
+        //Trajectory
+        cav_msgs::Trajectory trajectory;
+        //get earth to map tf
+        try
+        {
+            geometry_msgs::TransformStamped tf = tf2_buffer_.lookupTransform("earth", "map", ros::Time(0));
+            trajectory = TrajectoryPlantoTrajectory(trajectory_plan, tf);
+        }
+        catch (tf2::TransformException &ex)
+        {
+            //Throw exception
+            ROS_WARN("%s", ex.what());
+        }
+        request_msg.trajectory = trajectory;
+        request_msg.expiration = trajectory_plan.back().target_time.toSec();
+        
+        return request_msg;
+    }
+
+    cav_msgs::Trajectory CooperativeLaneChangePlugin::TrajectoryPlantoTrajectory(const std::vector<cav_msgs::TrajectoryPlanPoint>& traj_points, const geometry_msgs::TransformStamped& tf) const{
+        cav_msgs::Trajectory traj;
+        cav_msgs::LocationECEF ecef_location = TrajectoryPointtoECEF(traj_points[0], tf);
+
+        if (traj_points.size()<2){
+            ROS_WARN("Received Trajectory Plan is too small");
+            traj.offsets = {};
+        }
+        else{
+            for (size_t i=1; i<traj_points.size(); i++){
```

Comment:
I want to double check that we're not missing an implicit requirement here: 
```
# This message uses one start location in ECEF frame and a list of offset
# points relative to X, Y and Z of start point to represent a trajectory.
# The time interval between each pair of points should be 0.1 seconds.
```
Are we guaranteeing that time interval between each of these points?

Reply:
Would this mean that the interval should be atleast 0.1s? or exactly 0.1s?

Reply:
Should be exactly 0.1s, within reasonable assumptions about constant acceleration